File: tests/SnapBrief.Core.Tests/SessionModelTests.cs
using SnapBrief.Core.Editing;
using SnapBrief.Core.Exporting;
using SnapBrief.Core.Models;

namespace SnapBrief.Core.Tests;

public sealed class SessionModelTests
{
    private static readonly DateTimeOffset Start = new(2026, 9, 8, 20, 0, 0, TimeSpan.Zero);

    [Fact]
    public void Reorder_remove_and_undo_preserve_capture_identity_and_labels_follow_order()
    {
        var first = CaptureItem.Create("source/first.png", 100, 100, title: "Первый");
        var second = CaptureItem.Create("source/second.png", 100, 100, title: "Второй");
        var session = SessionOperations.AddCapture(SnapBriefSession.Create(Start), first, Start.AddMinutes(1));
        session = SessionOperations.AddCapture(session, second, Start.AddMinutes(2));
        var history = new SessionHistory(session, new FrozenTimeProvider(Start.AddMinutes(10)));

        history.Apply(value => SessionOperations.MoveCapture(value, second.Id, 0, Start.AddMinutes(3)));
        history.Apply(value => SessionOperations.RemoveCapture(value, first.Id, Start.AddMinutes(4)));

        Assert.Single(history.Current.Captures);
        Assert.Equal(second.Id, history.Current.Captures[0].Id);
        Assert.True(history.Undo());
        Assert.Equal([second.Id, first.Id], history.Current.Captures.Select(capture => capture.Id));
        Assert.Equal(5, history.Current.Revision);

        var prompt = new PromptGenerator().Generate(history.Current);
        Assert.Contains("Снимок A — Второй.", prompt, StringComparison.Ordinal);
        Assert.Contains("Снимок B — Первый.", prompt, StringComparison.Ordinal);
    }

    [Fact]
    public void Prompt_preserves_unicode_line_breaks_and_skips_empty_annotation_notes()
    {
        var capture = CaptureItem.Create("source/a.png", 100, 100, note: "Строка 1\nСтрока 2 🙂") with
        {
            Annotations =
            [
                AnnotationItem.Create(AnnotationKind.Arrow, [new(0.1, 0.2), new(0.8, 0.9)], note: "Сделать шире"),
                AnnotationItem.Create(AnnotationKind.Rectangle, [new(0.2, 0.2), new(0.4, 0.4)])
            ]
        };
        var session = SessionOperations.AddCapture(SnapBriefSession.Create(Start), capture, Start) with
        {
            GlobalNote = "Сохранить цвета"
        };

        var prompt = new PromptGenerator().Generate(session);

        Assert.Contains("Сохранить цвета", prompt, StringComparison.Ordinal);
        Assert.Contains("Строка 1\nСтрока 2 🙂", prompt, StringComparison.Ordinal);
        Assert.Contains("A1: Сделать шире", prompt, StringComparison.Ordinal);
        Assert.DoesNotContain("A2:", prompt, StringComparison.Ordinal);
        Assert.Equal(
            ["A1"],
            CaptureLabels.ForNotedAnnotations("A", capture).Select(item => item.DisplayLabel));
    }

    [Fact]
    public void Captures_without_notes_get_no_section_and_a_silent_session_gets_no_text()
    {
        var silent = CaptureItem.Create("source/a.png", 100, 100) with
        {
            Annotations = [AnnotationItem.Create(AnnotationKind.Rectangle, [new(0.2, 0.2), new(0.4, 0.4)])]
        };
        var noted = CaptureItem.Create("source/b.png", 100, 100) with
        {
            Annotations = [AnnotationItem.Create(AnnotationKind.Arrow, [new(0.1, 0.2), new(0.8, 0.9)], note: "Сделать шире")]
        };
        var silentSession = SessionOperations.AddCapture(SnapBriefSession.Create(Start), silent, Start);
        var mixedSession = SessionOperations.AddCapture(silentSession, noted, Start.AddSeconds(1));

        Assert.Equal(string.Empty, new PromptGenerator().Generate(silentSession));

        var prompt = new PromptGenerator().Generate(mixedSession);

        Assert.DoesNotContain("Снимок A", prompt, StringComparison.Ordinal);
        Assert.Contains("Снимок B.", prompt, StringComparison.Ordinal);
        Assert.Contains("B1: Сделать шире", prompt, StringComparison.Ordinal);
        Assert.Equal(
            "Общее пожелание:\nСохранить цвета",
            new PromptGenerator().Generate(silentSession with { GlobalNote = "Сохранить цвета" }));
    }

    [Fact]
    public void Sent_captures_leave_the_package_and_give_their_letters_to_the_waiting_ones()
    {
        var first = CaptureItem.Create("source/first.png", 100, 100) with { Sent = true };
        var second = CaptureItem.Create("source/second.png", 100, 100);
        var third = CaptureItem.Create("source/third.png", 100, 100) with { Sent = true };
        var fourth = CaptureItem.Create("source/fourth.png", 100, 100);
        CaptureItem[] captures = [first, second, third, fourth];

        var package = SentCaptureRules.ForPackage(captures, capture => capture.Sent);
        var labels = SentCaptureRules.StripLabels([.. captures.Select(capture => capture.Sent)]);

        Assert.Equal([second.Id, fourth.Id], package.Select(capture => capture.Id));
        Assert.Equal([null, "A", null, "B"], labels);
        Assert.Equal(["A", "B"], SentCaptureRules.StripLabels([false, false]));
        Assert.Empty(SentCaptureRules.ForPackage(captures, _ => true));
    }

    private sealed class FrozenTimeProvider(DateTimeOffset value) : TimeProvider
    {
        public override DateTimeOffset GetUtcNow() => value;
    }
}

File: macos/Tests/SnapikCoreTests/SessionModelTests.swift
import XCTest

@testable import SnapikCore

/// Port of `tests/Snapik.Core.Tests/SessionModelTests.cs`.
final class SessionModelTests: XCTestCase {
    private static let start = ISO8601Precise.makeUTC(year: 2026, month: 9, day: 8, hour: 20, minute: 0, second: 0)

    private struct FrozenTimeProvider: TimeProvider {
        let value: Date
        func utcNow() -> Date { value }
    }

    func test_Reorder_remove_and_undo_preserve_capture_identity_and_labels_follow_order() throws {
        let start = Self.start
        let first = CaptureItem.create(sourceImagePath: "source/first.png", pixelWidth: 100, pixelHeight: 100, title: "Первый")
        let second = CaptureItem.create(sourceImagePath: "source/second.png", pixelWidth: 100, pixelHeight: 100, title: "Второй")
        var session = try SessionOperations.addCapture(
            SnapikSession.create(nowUtc: start), capture: first, nowUtc: start.addingTimeInterval(60))
        session = try SessionOperations.addCapture(session, capture: second, nowUtc: start.addingTimeInterval(120))
        let history = SessionHistory(initial: session, timeProvider: FrozenTimeProvider(value: start.addingTimeInterval(600)))

        try history.apply { value in try SessionOperations.moveCapture(value, captureId: second.id, destinationIndex: 0, nowUtc: start.addingTimeInterval(180)) }
        try history.apply { value in try SessionOperations.removeCapture(value, captureId: first.id, nowUtc: start.addingTimeInterval(240)) }

        XCTAssertEqual(1, history.current.captures.count)
        XCTAssertEqual(second.id, history.current.captures[0].id)
        XCTAssertTrue(history.undo())
        XCTAssertEqual([second.id, first.id], history.current.captures.map { $0.id })
        XCTAssertEqual(5, history.current.revision)

        let prompt = try PromptGenerator().generate(history.current)
        XCTAssertTrue(prompt.contains("Снимок A — Второй."))
        XCTAssertTrue(prompt.contains("Снимок B — Первый."))
    }

    func test_Prompt_preserves_unicode_line_breaks_and_skips_empty_annotation_notes() throws {
        let start = Self.start
        var capture = CaptureItem.create(
            sourceImagePath: "source/a.png", pixelWidth: 100, pixelHeight: 100, note: "Строка 1\nСтрока 2 🙂")
        capture.annotations = [
            AnnotationItem.create(kind: .arrow, points: [NormalizedPoint(0.1, 0.2), NormalizedPoint(0.8, 0.9)], note: "Сделать шире"),
            AnnotationItem.create(kind: .rectangle, points: [NormalizedPoint(0.2, 0.2), NormalizedPoint(0.4, 0.4)]),
        ]
        var session = try SessionOperations.addCapture(SnapikSession.create(nowUtc: start), capture: capture, nowUtc: start)
        session.globalNote = "Сохранить цвета"

        let prompt = try PromptGenerator().generate(session)

        XCTAssertTrue(prompt.contains("Сохранить цвета"))
        XCTAssertTrue(prompt.contains("Строка 1\nСтрока 2 🙂"))
        XCTAssertTrue(prompt.contains("A1: Сделать шире"))
        XCTAssertFalse(prompt.contains("A2:"))
        XCTAssertEqual(
            ["A1"],
            CaptureLabels.forNotedAnnotations(captureLabel: "A", capture: capture).map { $0.displayLabel })
    }

    /// Port of `SessionModelTests.Captures_without_notes_get_no_section_and_a_silent_session_gets_no_text`.
    func test_Captures_without_notes_get_no_section_and_a_silent_session_gets_no_text() throws {
        let start = Self.start
        var silent = CaptureItem.create(sourceImagePath: "source/a.png", pixelWidth: 100, pixelHeight: 100)
        silent.annotations = [
            AnnotationItem.create(kind: .rectangle, points: [NormalizedPoint(0.2, 0.2), NormalizedPoint(0.4, 0.4)])
        ]
        var noted = CaptureItem.create(sourceImagePath: "source/b.png", pixelWidth: 100, pixelHeight: 100)
        noted.annotations = [
            AnnotationItem.create(
                kind: .arrow, points: [NormalizedPoint(0.1, 0.2), NormalizedPoint(0.8, 0.9)], note: "Сделать шире")
        ]
        let silentSession = try SessionOperations.addCapture(
            SnapikSession.create(nowUtc: start), capture: silent, nowUtc: start)
        let mixedSession = try SessionOperations.addCapture(
            silentSession, capture: noted, nowUtc: start.addingTimeInterval(1))

        XCTAssertEqual("", try PromptGenerator().generate(silentSession))

        let prompt = try PromptGenerator().generate(mixedSession)

        XCTAssertFalse(prompt.contains("Снимок A"))
        XCTAssertTrue(prompt.contains("Снимок B."))
        XCTAssertTrue(prompt.contains("B1: Сделать шире"))

        var withGlobalNote = silentSession
        withGlobalNote.globalNote = "Сохранить цвета"
        XCTAssertEqual(
            "Общее пожелание:\nСохранить цвета", try PromptGenerator().generate(withGlobalNote))
    }

    /// Port of `SessionModelTests.Sent_captures_leave_the_package_and_give_their_letters_to_the_waiting_ones`.
    func test_Sent_captures_leave_the_package_and_give_their_letters_to_the_waiting_ones() throws {
        var first = CaptureItem.create(sourceImagePath: "source/first.png", pixelWidth: 100, pixelHeight: 100)
        first.sent = true
        let second = CaptureItem.create(sourceImagePath: "source/second.png", pixelWidth: 100, pixelHeight: 100)
        var third = CaptureItem.create(sourceImagePath: "source/third.png", pixelWidth: 100, pixelHeight: 100)
        third.sent = true
        let fourth = CaptureItem.create(sourceImagePath: "source/fourth.png", pixelWidth: 100, pixelHeight: 100)
        let captures = [first, second, third, fourth]

        let package = SentCaptureRules.forPackage(captures) { $0.sent }
        let labels = try SentCaptureRules.stripLabels(captures.map { $0.sent })

        XCTAssertEqual([second.id, fourth.id], package.map { $0.id })
        XCTAssertEqual([nil, "A", nil, "B"], labels)
        XCTAssertEqual(["A", "B"] as [String?], try SentCaptureRules.stripLabels([false, false]))
        XCTAssertTrue(SentCaptureRules.forPackage(captures) { _ in true }.isEmpty)
    }

    /// Port of `SessionModelTests.A_full_strip_stops_at_twenty_six_captures_and_its_last_letter_is_Z`.
    /// The limit of the strip and its letters are one rule: a full strip asks for the letter Z and
    /// never for the one after it. The soft warning is below the limit, or it would never be said.
    func test_A_full_strip_stops_at_twenty_six_captures_and_its_last_letter_is_Z() throws {
        XCTAssertEqual(26, SentCaptureRules.maxStripCaptures)
        XCTAssertTrue(SentCaptureRules.softStripWarning < SentCaptureRules.maxStripCaptures)

        let labels = try SentCaptureRules.stripLabels(
            Array(repeating: false, count: SentCaptureRules.maxStripCaptures))

        XCTAssertEqual(SentCaptureRules.maxStripCaptures, labels.count)
        XCTAssertEqual("A", labels.first ?? nil)
        XCTAssertEqual("Z", labels.last ?? nil)
    }

}
